### PR TITLE
added the initial version of the (t *Torrent) KnownSwarm() function

### PR DESCRIPTION
**Pull request by boramalper**
For #183. Please check `TODO`s before/after merging.

### CODE REVIEW

File: torrent.go [hunk 1]
```diff
@@ -122,6 +123,44 @@ func (t *Torrent) Closed() <-chan struct{} {
...
+			Port: port,
+			Source: conn.Discovery,
+			// TODO: the connection can be unencrypted due to our (or the peer's) preference,
+			//       but the remote peer might support the encryption. Find a better way to query
+			//       that information, if possible.
+			SupportsEncryption: conn.encrypted,
+		})
+	}
+
+	// TODO: how can we add half-open peers?
```
Comment (anacrolix) [inline torrent.go]: You can get the addresses from Torrent.halfOpen.

File: torrent.go [hunk 2]
```diff
@@ -122,6 +123,44 @@ func (t *Torrent) Closed() <-chan struct{} {
...
+		if err != nil {
+			panic(err)
+		}
+
+		ks = append(ks, Peer{
+			Id: conn.PeerID,
+			IP: ip,
+			Port: port,
+			Source: conn.Discovery,
+			// TODO: the connection can be unencrypted due to our (or the peer's) preference,
```
Comment (anacrolix) [inline torrent.go]: SupportsEncryption is advisory only. I don't believe anything actually acts on it, but the information is provided by some sources.

Comment (boramalper) [reply torrent.go]: > but the information is provided by some sources

In this case, by which sources? `conn.encrypted` is enough or do I have another option?

File: torrent.go (hunk 1 above)
Comment (boramalper) [reply torrent.go]: But how do I get their peer ID, *source*, and whether or not they support encryption? `Torrent.halfOpen` just have the hostport strings right?

Comment (anacrolix) [reply torrent.go]: That's right. I don't think we can as it is. We could however carry over the values when they're pulled from the pending peers list, and put it in the half-open map. Once the connection is successful, we can apply them there.

Comment (anacrolix) [reply torrent.go]: See b39df82fbc9269c69257ac22dea87338a886faf6

File: torrent.go (hunk 2 above)
Comment (anacrolix) [reply torrent.go]: If the connection is encrypted, that's certainly enough to set SupportsEncryption. But if we're not connected to them with an encrypted connection, I couldn't say what's appropriate. We can carry forward the SupportsEncryption value as we received it from trackers/DHT/PEX, or just use the encryption state for the connection. It's probably easiest to do the latter for now.

**Pull request merged**